piet-cairo/src/text.rs

```diff
@@ -210,235 +357,149 @@ impl TextLayout for CairoTextLayout {
+            let pango_width = new_width * pango::SCALE as f64;
+            self.pango_layout.set_width(pango_width as i32);
+        } else {
+            /*
+             * NOTE: -1 is the default value, however `update_width` *could*
+             * be called any number of times with different values so we need
+             * to make sure to reset back to default whenever we get no width
+             */
+            self.pango_layout.set_width(-1);
+        }
 
-    // Now that we know it's not beginning or end, begin binary search.
-    // Iterative style
-    let mut left = start;
-    let mut right = end;
-    loop {
-        // pick halfway point
-        let middle = left + ((right - left) / 2);
-
-        let grapheme_bounds = match get_grapheme_boundaries(font, text, middle) {
-            Some(bounds) => bounds,
-            None => return HitTestPoint::default(),
-        };
+        let mut line_metrics = Vec::new();
+        let mut y_distance = 0.;
+        let mut widest_logical_width = 0;
+        for line_index in 0..self.pango_layout.get_line_count() {
+            if let Some(line) = self.pango_layout.get_line_readonly(line_index) {
+                /*
+                 * NOTE: These values are not currently exposed so we need to get them
+                 * manually. It's kinda sucky I know
+                 *
+                 * TODO(ForLoveOfCats): Submit a PR to gtk-rs to expose these values
+                 */
+                let (start_offset, end_offset) = unsafe {
+                    let raw_line = line.to_glib_none();
+
+                    let start_offset = (*raw_line.0).start_index as usize;
+                    let length = (*raw_line.0).length as usize;
+
+                    (start_offset, start_offset + length)
+                };
 
-        if let Some(hit) = point_x_in_grapheme(point.x, &grapheme_bounds) {
-            return hit;
-        }
+                //Pango likes to give us the line range *without* the newline char(s)
+                let end_offset = {
+                    let mut end_offset = end_offset;
+                    loop {
+                        let next_byte = self.text.as_bytes().get(end_offset);
+                        match next_byte {
+                            Some(b'\r') | Some(b'\n') => {
+                                end_offset += 1;
+                            }
+
+                            _ => break end_offset,
+                        }
+                    }
+                };
 
-        // since it's not a hit, check if closer to start or finish
-        // and move the appropriate search boundary
-        if point.x < grapheme_bounds.leading {
-            right = middle;
-        } else if point.x > grapheme_bounds.trailing {
-            left = middle + 1;
-        } else {
-            unreachable!("hit_test_point conditional is exhaustive");
+                let ink_rect = line.get_extents().0;
+                let logical_rect = line.get_extents().1;
```

Comment:
to clarify these two things: the logical rect is going to be the ascent + the descent (and maybe also + leading) as the y value, and then the sum of the advances of the runs as the x value.  The *ink rect* is going to be the bounding box of the actual drawn glyphs; this can be different because glyphs can extend outside of their logical bounds.

In any case, neither of these are related to the baseline; for that we need to use the `PangoLayoutIter` interface. :)

Reply:
My logic was that because the Pango origin is in the upper left corner that adding the ink height to the top of the line would give the bottom line of the glyphs which is what understood baseline to be. Good to know that Pango has a mechanism for handling this calculation directly :)

Reply:
Fair enough. The way to think about 'baseline' is that it's the line you would be writing on if you were writing by hand in a ruled notebook. The descenders (on g, y, p, etc.) go below the baseline, and everything else is above it.

Reply:
I've updated it to use the baseline information provided by `LayoutIter` but I'm not seeing any noticable change in snapshot 11 as a result.

Reply:
Looking at it with my untrained eye shows some stuff wrong from the same behavior causing `eol_hit_testing` to fail but the outside hits on the bottom are looking very wrong :thinking:

Reply:
Yes it looks like something is still funky with the baseline calculations, I'll pay around a bit...

Reply:
From the docs, it looks like [`get_baseline`](https://developer.gnome.org/pygtk/unstable/class-pangolayoutiter.html#method-pangolayoutiter--get-baseline) returns an offset relative to the origin of the whole layout, not the specific line; our `LineMetrics` struct expects this value to be relative to the `y_offset` of the specific line, so you'll need to subtract that to get the correct value.

Reply:
Ah ha, that explains why these results are visually identical to the old results where I was manually adding the y_offset